node/core/provisioner/src/lib.rs

```diff
@@ -0,0 +1,362 @@
+		}
+	}
+}
+
+impl TryFrom<AllMessages> for ToJob {
+	type Error = ();
+
+	fn try_from(msg: AllMessages) -> Result<Self, Self::Error> {
+		match msg {
+			AllMessages::Provisioner(pm) => Ok(Self::Provisioner(pm)),
+			_ => Err(()),
+		}
+	}
+}
+
+impl From<ProvisionerMessage> for ToJob {
+	fn from(pm: ProvisionerMessage) -> Self {
+		Self::Provisioner(pm)
+	}
+}
+
+pub enum FromJob {
+	Runtime(RuntimeApiMessage),
+}
+
+impl From<FromJob> for AllMessages {
+	fn from(from_job: FromJob) -> AllMessages {
+		match from_job {
+			FromJob::Runtime(ram) => AllMessages::RuntimeApi(ram),
+		}
+	}
+}
+
+impl TryFrom<AllMessages> for FromJob {
+	type Error = ();
+
+	fn try_from(msg: AllMessages) -> Result<Self, Self::Error> {
+		match msg {
+			AllMessages::RuntimeApi(runtime) => Ok(FromJob::Runtime(runtime)),
+			_ => Err(()),
+		}
+	}
+}
+
+#[derive(Debug, derive_more::From)]
+pub enum Error {
+	#[from]
+	Sending(mpsc::SendError),
+	#[from]
+	Util(util::Error),
+	#[from]
+	OneshotRecv(oneshot::Canceled),
+	OneshotSend,
+}
+
+impl JobTrait for ProvisioningJob {
+	type ToJob = ToJob;
+	type FromJob = FromJob;
+	type Error = Error;
+	type RunArgs = ();
+
+	const NAME: &'static str = "ProvisioningJob";
+
+	/// Run a job for the parent block indicated
+	//
+	// this function is in charge of creating and executing the job's main loop
+	fn run(
+		relay_parent: Hash,
+		_run_args: Self::RunArgs,
+		receiver: mpsc::Receiver<ToJob>,
+		sender: mpsc::Sender<FromJob>,
+	) -> Pin<Box<dyn Future<Output = Result<(), Self::Error>> + Send>> {
+		async move {
+			let job = ProvisioningJob::new(relay_parent, sender, receiver);
+
+			// it isn't necessary to break run_loop into its own function,
+			// but it's convenient to separate the concerns in this way
+			job.run_loop().await
+		}
+		.boxed()
+	}
+}
+
+impl ProvisioningJob {
+	pub fn new(
+		relay_parent: Hash,
+		sender: mpsc::Sender<FromJob>,
+		receiver: mpsc::Receiver<ToJob>,
+	) -> Self {
+		Self {
+			relay_parent,
+			sender,
+			receiver,
+			provisionable_data_channels: Vec::new(),
+			backed_candidates: Vec::new(),
+			signed_bitfields: Vec::new(),
+		}
+	}
+
+	async fn run_loop(mut self) -> Result<(), Error> {
+		while let Some(msg) = self.receiver.next().await {
+			use ProvisionerMessage::{
+				ProvisionableData, RequestBlockAuthorshipData, RequestInherentData,
+			};
+
+			match msg {
+				ToJob::Provisioner(RequestInherentData(_, sender)) => {
+					// Note the cloning here: we have to clone the vectors of signed bitfields and backed candidates
+					// so that we can respond to more than a single request for inherent data; we can't just move them.
+					// It would be legal, however, to set up `from_job` as `&mut _` instead of a clone. We clone it instead
+					// of borrowing it so that this async function doesn't depend at all on the lifetime of `&self`.
+					send_inherent_data(
+						self.relay_parent,
+						self.signed_bitfields.clone(),
+						self.backed_candidates.clone(),
```

Comment:
I may misunderstand the comment here. My understanding is that borrowing `self` is not a problem unless we spawn a separate task to do this `send_inherent_data` work. In that case, it's probably more efficient to borrow `&self.signed_bitfields` and `&self.backed_candidates` and clone only the ones which get used.

However, the runtime of this function is not likely to be high. The worst case here is hundreds of backed candidates that each wrap a code upgrade of 10MB, so this clone might actually be significant.